File: src/main/scala/com/martomate/tripaint/util/CachedLoader.scala
package com.martomate.tripaint.util

import scala.util.{Failure, Success, Try}

object CachedLoader {

  /** @return `Success((value, foundInCache))` or `Failure(errorDuringLoad)` */
  def apply[T](cached: => Option[T], load: => Try[T]): Try[(T, Boolean)] =
    cached match
      case Some(value) => Success((value, true))
      case None =>
        load match
          case Success(value) => Success((value, false))
          case Failure(error) => Failure(error)
}

File: core/src/main/kotlin/tripaint/util/CachedLoader.kt
package tripaint.util

object CachedLoader {

    /** @return `Success((value, foundInCache))` or `Failure(errorDuringLoad)` */
    fun <T> apply(cached: () -> T?, load: () -> Result<T>): Result<Pair<T, Boolean>> {
        val cachedValue = cached()
        return if (cachedValue != null) {
            Result.success(Pair(cachedValue, true))
        } else {
            load().map { Pair(it, false) }
        }
    }
}
